dll/win32/shell32/CSendToMenu.cpp

```diff
@@ -176,114 +157,81 @@ void CSendToMenu::UnloadAllItems()
 {
     UnloadAllItems();
 
-    LPITEMIDLIST pidlSendTo;
-    m_pSendTo = GetSpecialFolder(hwnd, CSIDL_SENDTO, &pidlSendTo);
-    if (!m_pSendTo)
-    {
-        ERR("GetSpecialFolder\n");
-        return FALSE;
-    }
+    CComHeapPtr<ITEMIDLIST_ABSOLUTE> pidlSendTo;
+    HRESULT hr = GetSpecialFolder(hwnd, &m_pSendTo, CSIDL_SENDTO, &pidlSendTo);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
 
-    HRESULT hr;
     CComPtr<IEnumIDList> pEnumIDList;
     hr = m_pSendTo->EnumObjects(hwnd,
                                 SHCONTF_FOLDERS | SHCONTF_NONFOLDERS,
                                 &pEnumIDList);
-    if (FAILED(hr))
-    {
-        ERR("EnumObjects: %08lX\n", hr);
-        ILFree(pidlSendTo);
-        return FALSE;
-    }
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
 
-    BOOL bOK = TRUE;
-    LPITEMIDLIST pidlChild;
-    UINT nCount = 0;
-    while (pEnumIDList->Next(1, &pidlChild, NULL) == S_OK)
+    hr = S_OK;
+    PITEMID_CHILD child;
+    while (pEnumIDList->Next(1, &child, NULL) == S_OK)
     {
-        SENDTO_ITEM *pNewItem = (SENDTO_ITEM *)HeapAlloc(GetProcessHeap(),
-                                                         HEAP_ZERO_MEMORY,
-                                                         sizeof(SENDTO_ITEM));
-        if (!pNewItem)
-        {
-            ERR("HeapAlloc\n");
-            bOK = FALSE;
-            CoTaskMemFree(pidlChild);
-            break;
-        }
+        CComHeapPtr<ITEMID_CHILD> pidlChild(child);
 
         STRRET strret;
         hr = m_pSendTo->GetDisplayNameOf(pidlChild, SHGDN_NORMAL, &strret);
-        if (SUCCEEDED(hr))
+        if (FAILED_UNEXPECTEDLY(hr))
+            continue;
+
+        CComHeapPtr<OLECHAR> str;
+        if (strret.uType == STRRET_WSTR)
         {
-            LPWSTR pszText = NULL;
-            hr = StrRetToStrW(&strret, pidlChild, &pszText);
-            if (SUCCEEDED(hr))
-            {
-                LPITEMIDLIST pidlAbsolute = ILCombine(pidlSendTo, pidlChild);
-
-                SHFILEINFOW fi = { NULL };
-                const UINT uFlags = SHGFI_PIDL | SHGFI_TYPENAME |
-                                    SHGFI_ICON | SHGFI_SMALLICON;
-                SHGetFileInfoW((LPCWSTR)pidlAbsolute, 0, &fi, sizeof(fi), uFlags);
-
-                ILFree(pidlAbsolute);
-
-                pNewItem->pidlChild = pidlChild;
-                pNewItem->pszText = pszText;
-                pNewItem->hIcon = fi.hIcon;
-                if (m_pItems)
-                {
-                    pNewItem->pNext = m_pItems;
-                }
-                m_pItems = pNewItem;
-
-                // successful
-                ++nCount;
-                if (nCount >= MAX_ITEM_COUNT)
-                {
-                    break;
-                }
-                continue;
-            }
-            else
-            {
-                ERR("StrRetToStrW: %08lX\n", hr);
-            }
+            str.Attach(strret.pOleStr);
```

Comment:
This will result in a double free. `StrRetToStrW` frees the `STRRET` so there's no need to do it a second time.